File: DFS/LetterCombination/letterCombinations.py
class Solution:
    def letterCombinations(self, digits: str) -> list[str]:
        """
        state = {}
        for 每个数字
            for 当前数字的备选字母
                for state中的所有字符串
        """
        if not digits:
            return []

        char = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        res = [""]
        for item in digits:
            now = []
            for c in char[item]:
                for u in res:
                    now.append(c + u)

            res = now

        return res
        

File: DFS/LetterCombination/letterCombinations.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <queue>
#include <cstdio>

using namespace std;

class Solution
{
public:
    string wordDict[8] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    vector<string> letterCombinations(string digits)
    {
        if (digits.empty())
            return vector<string>();

        vector<string> res(1, "");
        for (auto item : digits)
        {
            vector<string> now;
            for (auto c : wordDict[item - '2'])
                for (auto u : res)
                    now.push_back(u + c);

            res = now;
        }

        return res;
    }
};